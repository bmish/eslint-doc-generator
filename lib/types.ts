import type { RuleDocTitleFormat } from './rule-doc-title-format.js';
import type { TSESLint, JSONSchema } from '@typescript-eslint/utils';

// Standard ESLint types.

export type RuleModule = TSESLint.RuleModule<string, readonly unknown[]>;

export type Rules = TSESLint.Linter.RulesRecord;

export type RuleSeverity = TSESLint.Linter.RuleLevel;

export type Config = TSESLint.Linter.Config;

export type Plugin = TSESLint.Linter.Plugin;

// Custom types.

export const SEVERITY_ERROR = new Set<RuleSeverity>([2, 'error']);
export const SEVERITY_WARN = new Set<RuleSeverity>([1, 'warn']);
export const SEVERITY_OFF = new Set<RuleSeverity>([0, 'off']);

export enum SEVERITY_TYPE {
  'error' = 'error',
  'warn' = 'warn',
  'off' = 'off',
}

export const SEVERITY_TYPE_TO_SET: {
  [key in SEVERITY_TYPE]: Set<TSESLint.Linter.RuleLevel>;
} = {
  [SEVERITY_TYPE.error]: SEVERITY_ERROR,
  [SEVERITY_TYPE.warn]: SEVERITY_WARN,
  [SEVERITY_TYPE.off]: SEVERITY_OFF,
};

export type ConfigsToRules = Record<string, Rules>;

export interface RuleDetails {
  name: string;
  description?: string; // Rule might not have a description.
  fixable: boolean;
  hasSuggestions: boolean;
  requiresTypeChecking: boolean;
  deprecated: boolean;
  schema: JSONSchema.JSONSchema4;
  type?: string; // Rule might not have a type.
}

/**
 * Some configs may have an emoji defined.
 */
export type ConfigEmojis = readonly { config: string; emoji: string }[];

/**
 * Rule doc notices.
 */
export enum NOTICE_TYPE {
  CONFIGS = 'configs',
  DEPRECATED = 'deprecated',
  FIXABLE = 'fixable',
  FIXABLE_AND_HAS_SUGGESTIONS = 'fixableAndHasSuggestions', // Consolidated notice for space-saving.
  HAS_SUGGESTIONS = 'hasSuggestions',
  OPTIONS = 'options',
  REQUIRES_TYPE_CHECKING = 'requiresTypeChecking',
  TYPE = 'type',
}

/**
 * Rule list columns.
 */
export enum COLUMN_TYPE {
  CONFIGS_ERROR = 'configsError',
  CONFIGS_OFF = 'configsOff',
  CONFIGS_WARN = 'configsWarn',
  DEPRECATED = 'deprecated',
  DESCRIPTION = 'description',
  FIXABLE = 'fixable',
  FIXABLE_AND_HAS_SUGGESTIONS = 'fixableAndHasSuggestions', // Consolidated column for space-saving.
  HAS_SUGGESTIONS = 'hasSuggestions',
  NAME = 'name',
  OPTIONS = 'options',
  REQUIRES_TYPE_CHECKING = 'requiresTypeChecking',
  TYPE = 'type',
}

/**
 * CLI/config file options.
 */
export enum OPTION_TYPE {
  CHECK = 'check',
  CONFIG_EMOJI = 'configEmoji',
  IGNORE_CONFIG = 'ignoreConfig',
  IGNORE_DEPRECATED_RULES = 'ignoreDeprecatedRules',
  INIT_RULE_DOCS = 'initRuleDocs',
  PATH_RULE_DOC = 'pathRuleDoc',
  PATH_RULE_LIST = 'pathRuleList',
  POSTPROCESS = 'postprocess',
  RULE_DOC_NOTICES = 'ruleDocNotices',
  RULE_DOC_SECTION_EXCLUDE = 'ruleDocSectionExclude',
  RULE_DOC_SECTION_INCLUDE = 'ruleDocSectionInclude',
  RULE_DOC_SECTION_OPTIONS = 'ruleDocSectionOptions',
  RULE_DOC_TITLE_FORMAT = 'ruleDocTitleFormat',
  RULE_LIST_COLUMNS = 'ruleListColumns',
  RULE_LIST_SPLIT = 'ruleListSplit',
  URL_CONFIGS = 'urlConfigs',
  URL_RULE_DOC = 'urlRuleDoc',
}

// JSDocs for options should be kept in sync with README.md and the CLI runner in cli.ts.
/** The type for the config file (e.g. `.eslint-doc-generatorrc.js`) and internal `generate()` function. */
export type GenerateOptions = {
  /** Whether to check for and fail if there is a diff. No output will be written. Typically used during CI. Default: `false`. */
  readonly check?: boolean;
  /**
   * List of configs and their associated emojis.
   * Array of `[configName, emoji]`.
   * Default emojis are provided for common configs.
   * To remove a default emoji and rely on a badge instead, provide the config name without an emoji.
   */
  readonly configEmoji?: readonly (readonly string[])[];
  /** Configs to ignore from being displayed. Often used for an `all` config. */
  readonly ignoreConfig?: readonly string[];
  /** Whether to ignore deprecated rules from being checked, displayed, or updated. Default: `false`. */
  readonly ignoreDeprecatedRules?: boolean;
  /** Whether to create rule doc files if they don't yet exist. Default: `false`. */
  readonly initRuleDocs?: boolean;
  /** Path to markdown file for each rule doc. Use `{name}` placeholder for the rule name. Default: `docs/rules/{name}.md`. */
  readonly pathRuleDoc?: string;
  /** Path to markdown file(s) where the rules table list should live. Default: `README.md`. */
  readonly pathRuleList?: string | readonly string[];
  /**
   * Function to be called with the generated content and file path for each processed file.
   * Useful for applying custom transformations such as formatting with tools like prettier.
   * Only available via a JavaScript config file.
   */
  readonly postprocess?: (
    content: string,
    pathToFile: string
  ) => string | Promise<string>;
  /**
   * Ordered list of notices to display in rule doc.
   * Non-applicable notices will be hidden.
   * Choices: `configs`, `deprecated`, `fixable` (off by default), `fixableAndHasSuggestions`, `hasSuggestions` (off by default), `options` (off by default), `requiresTypeChecking`, `type` (off by default).
   * Default: `['deprecated', 'configs', 'fixableAndHasSuggestions', 'requiresTypeChecking']`.
   */
  readonly ruleDocNotices?: readonly NOTICE_TYPE[];
  /** Disallowed sections in each rule doc. Exit with failure if present. */
  readonly ruleDocSectionExclude?: readonly string[];
  /** Required sections in each rule doc. Exit with failure if missing. */
  readonly ruleDocSectionInclude?: readonly string[];
  /** Whether to require an "Options" or "Config" rule doc section and mention of any named options for rules with options. Default: `true`. */
  readonly ruleDocSectionOptions?: boolean;
  /** The format to use for rule doc titles. Default: `desc-parens-prefix-name`. */
  readonly ruleDocTitleFormat?: RuleDocTitleFormat;
  /**
   * Ordered list of columns to display in rule list.
   * Empty columns will be hidden.
   * Choices: `configsError`, `configsOff`, `configsWarn`, `deprecated`, `description`, `fixable`, `fixableAndHasSuggestions` (off by default), `hasSuggestions`, `name`, `options` (off by default), `requiresTypeChecking`, `type` (off by default).
   * Default: `['name', 'description', 'configsError', 'configsWarn', 'configsOff', 'fixable', 'hasSuggestions', 'requiresTypeChecking', 'deprecated']`.
   */
  readonly ruleListColumns?: readonly COLUMN_TYPE[];
  /**
   * Rule property to split the rules list by.
   * A separate list and header will be created for each value.
   * Example: `meta.type`.
   */
  readonly ruleListSplit?: string;
  /** Link to documentation about the ESLint configurations exported by the plugin. */
  readonly urlConfigs?: string;
  /**
   * Link to documentation for each rule.
   * Useful when it differs from the rule doc path on disk (e.g. custom documentation site in use).
   * Use `{name}` placeholder for the rule name.
   */
  readonly urlRuleDoc?: string;
};
