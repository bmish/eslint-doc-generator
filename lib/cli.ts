import { Command, Argument, Option } from 'commander';
import { RULE_DOC_TITLE_FORMATS } from './rule-doc-title-format.js';
import { OPTION_DEFAULTS } from './options.js';
import { cosmiconfig } from 'cosmiconfig';
import Ajv from 'ajv';
import merge from 'deepmerge';
import {
  COLUMN_TYPE,
  NOTICE_TYPE,
  GenerateOptions,
  OPTION_TYPE,
} from './types.js';
import { getCurrentPackageVersion } from './package-json.js';

/**
 * Used for collecting repeated CLI options into an array.
 * Example: --foo bar --foo baz => ['bar', 'baz']
 */
function collect(value: string, previous: string[]): string[] {
  return [...previous, value];
}

/**
 * Used for collecting CSV CLI options into an array.
 * Example: --foo bar,baz,buz => ['bar', 'baz', 'buz']
 * */
function collectCSV(value: string, previous: string[]): string[] {
  return [...previous, ...value.split(',')];
}

/**
 * Used for collecting repeated, nested CSV CLI options into an array of arrays.
 * Example: --foo baz,bar --foo biz,buz => [['baz', 'bar'], ['biz', 'buz']]
 * */
function collectCSVNested(value: string, previous: string[][]): string[][] {
  return [...previous, value.split(',')];
}

function parseBoolean(value: string | undefined): boolean {
  return ['true', undefined].includes(value);
}

/**
 * Load and validate the config file.
 * Cosmiconfig supports many possible filenames/formats.
 */
async function loadConfigFileOptions(): Promise<GenerateOptions> {
  const explorer = cosmiconfig('eslint-doc-generator');
  const explorerResults = await explorer.search();
  if (explorerResults && !explorerResults.isEmpty) {
    // Validate schema for config file.
    const schemaStringArray = {
      type: 'array',
      uniqueItems: true,
      minItems: 1,
      items: {
        type: 'string',
        minLength: 1,
      },
    };

    const schemaConfigEmoji = {
      // Top-level array.
      type: 'array',
      uniqueItems: true,
      minItems: 1,
      items: {
        // Nested array (config/emoji tuple).
        type: 'array',
        uniqueItems: true,
        minItems: 1, // Allowed to pass only config name to remove default emoji.
        maxItems: 2, // Normally, two items will be passed.
        items: {
          type: 'string',
          minLength: 1,
        },
      },
    };

    const properties: { [key in OPTION_TYPE]: unknown } = {
      check: { type: 'boolean' },
      configEmoji: schemaConfigEmoji,
      ignoreConfig: schemaStringArray,
      ignoreDeprecatedRules: { type: 'boolean' },
      initRuleDocs: { type: 'boolean' },
      pathRuleDoc: { type: 'string' },
      pathRuleList: { anyOf: [{ type: 'string' }, schemaStringArray] },
      postprocess: {
        /* JSON Schema can't validate functions so check this later */
      },
      ruleDocNotices: schemaStringArray,
      ruleDocSectionExclude: schemaStringArray,
      ruleDocSectionInclude: schemaStringArray,
      ruleDocSectionOptions: { type: 'boolean' },
      ruleDocTitleFormat: { type: 'string' },
      ruleListColumns: schemaStringArray,
      splitBy: { type: 'string' },
      urlConfigs: { type: 'string' },
      urlRuleDoc: { type: 'string' },
    };
    const schema = {
      type: 'object',
      properties,
      additionalProperties: false,
    };

    const ajv = new Ajv();
    const validate = ajv.compile(schema);
    const valid = validate(explorerResults.config);
    if (!valid) {
      throw new Error(
        validate.errors
          ? ajv.errorsText(validate.errors, { dataVar: 'config file' })
          : /* istanbul ignore next -- this shouldn't happen */
            'Invalid config file'
      );
    }

    if (
      explorerResults.config.postprocess &&
      typeof explorerResults.config.postprocess !== 'function'
    ) {
      throw new Error('postprocess must be a function');
    }

    return explorerResults.config;
  }
  return {};
}

/**
 * Run the CLI and gather options.
 * When this is done, load any config file and merge with the CLI options.
 * Finally, invoke a callback with the merged options.
 * Note: Does not introduce default values. Default values should be handled in the callback function.
 */
export async function run(
  argv: string[],
  cb: (path: string, options: GenerateOptions) => Promise<void>
) {
  const program = new Command();

  // Documentation for options should be kept in sync with README.md and the JSDocs for the `GenerateOptions` type.
  await program
    .version(getCurrentPackageVersion())
    .addArgument(
      new Argument('[path]', 'path to ESLint plugin root').default('.')
    )
    .option(
      '--check [boolean]',
      `(optional) Whether to check for and fail if there is a diff. No output will be written. Typically used during CI. (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.CHECK]
      })`,
      parseBoolean
    )
    .option(
      '--config-emoji <config-emoji>',
      '(optional) Custom emoji to use for a config. Format is `config-name,emoji`. Default emojis are provided for common configs. To remove a default emoji and rely on a badge instead, provide the config name without an emoji. Option can be repeated.',
      collectCSVNested,
      []
    )
    .option(
      '--ignore-config <config>',
      '(optional) Config to ignore from being displayed (often used for an `all` config) (option can be repeated).',
      collect,
      []
    )
    .option(
      '--ignore-deprecated-rules [boolean]',
      `(optional) Whether to ignore deprecated rules from being checked, displayed, or updated. (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.IGNORE_DEPRECATED_RULES]
      })`,
      parseBoolean
    )
    .option(
      '--init-rule-docs [boolean]',
      `(optional) Whether to create rule doc files if they don't yet exist. (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.INIT_RULE_DOCS]
      })`,
      parseBoolean
    )
    .option(
      '--path-rule-doc <path>',
      `(optional) Path to markdown file for each rule doc. Use \`{name}\` placeholder for the rule name. (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_DOC]
      })`
    )
    .option(
      '--path-rule-list <path>',
      `(optional) Path to markdown file where the rules table list should live. Option can be repeated. Defaults to ${
        OPTION_DEFAULTS[OPTION_TYPE.PATH_RULE_LIST]
      } if not provided.`,
      collect,
      []
    )
    .option(
      '--rule-doc-notices <notices>',
      `(optional) Ordered, comma-separated list of notices to display in rule doc. Non-applicable notices will be hidden. (choices: "${Object.values(
        NOTICE_TYPE
      ).join('", "')}") (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_NOTICES]
      })`,
      collectCSV,
      []
    )
    .option(
      '--rule-doc-section-exclude <section>',
      '(optional) Disallowed section in each rule doc (option can be repeated).',
      collect,
      []
    )
    .option(
      '--rule-doc-section-include <section>',
      '(optional) Required section in each rule doc (option can be repeated).',
      collect,
      []
    )
    .option(
      '--rule-doc-section-options [boolean]',
      `(optional) Whether to require an "Options" or "Config" rule doc section and mention of any named options for rules with options. (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_SECTION_OPTIONS]
      })`,
      parseBoolean
    )
    .addOption(
      new Option(
        '--rule-doc-title-format <format>',
        `(optional) The format to use for rule doc titles. (default: ${
          OPTION_DEFAULTS[OPTION_TYPE.RULE_DOC_TITLE_FORMAT]
        })`
      ).choices(RULE_DOC_TITLE_FORMATS)
    )
    .option(
      '--rule-list-columns <columns>',
      `(optional) Ordered, comma-separated list of columns to display in rule list. Empty columns will be hidden. (choices: "${Object.values(
        COLUMN_TYPE
      ).join('", "')})" (default: ${
        OPTION_DEFAULTS[OPTION_TYPE.RULE_LIST_COLUMNS]
      })`,
      collectCSV,
      []
    )
    .option(
      '--split-by <property>',
      '(optional) Rule property to split the rules list by. A separate list and header will be created for each value. Example: `meta.type`.'
    )
    .option(
      '--url-configs <url>',
      '(optional) Link to documentation about the ESLint configurations exported by the plugin.'
    )
    .option(
      '--url-rule-doc <url>',
      '(optional) Link to documentation for each rule. Useful when it differs from the rule doc path on disk (e.g. custom documentation site in use). Use `{name}` placeholder for the rule name.'
    )
    .action(async function (path, options: GenerateOptions) {
      // Load config file options and merge with CLI options.
      // CLI options take precedence.
      // For this to work, we can't have any default values from the CLI options that will override the config file options (except empty arrays, as arrays will be merged).
      // Default values should be handled in the callback function.
      const configFileOptions = await loadConfigFileOptions();

      // Perform any normalization needed ahead of merging.
      if (typeof configFileOptions.pathRuleList === 'string') {
        configFileOptions.pathRuleList = [configFileOptions.pathRuleList];
      }

      const generateOptions = merge(configFileOptions, options); // Recursive merge.

      // Invoke callback.
      await cb(path, generateOptions);
    })
    .parseAsync(argv);

  return program;
}
