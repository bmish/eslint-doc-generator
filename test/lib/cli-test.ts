import * as sinon from 'sinon';
import { run } from '../../lib/cli.js';
import mockFs from 'mock-fs';
import { OPTION_TYPE } from '../../lib/types.js';

const configFileOptionsAll: { [key in OPTION_TYPE]: unknown } = {
  check: true,
  configEmoji: ['recommended-from-config-file,ðŸš²'],
  ignoreConfig: [
    'ignoredConfigFromConfigFile1',
    'ignoredConfigFromConfigFile2',
  ],
  ignoreDeprecatedRules: true,
  initRuleDocs: true,
  pathRuleDoc: 'www.example.com/rule-doc-from-config-file',
  pathRuleList: 'www.example.com/rule-list-from-config-file',
  postprocess: (content: string) => content,
  ruleDocNotices: 'type',
  ruleDocSectionExclude: [
    'excludedSectionFromConfigFile1',
    'excludedSectionFromConfigFile2',
  ],
  ruleDocSectionInclude: [
    'includedSectionFromConfigFile1',
    'includedSectionFromConfigFile2',
  ],
  ruleDocSectionOptions: false,
  ruleDocTitleFormat: 'desc',
  ruleListColumns: 'fixable,hasSuggestions',
  splitBy: 'meta.docs.foo-from-config-file',
  urlConfigs: 'https://example.com/configs-url-from-config-file',
  urlRuleDoc: 'https://example.com/rule-doc-url-from-config-file',
};

const cliOptionsAll: { [key in OPTION_TYPE]: string[] } = {
  [OPTION_TYPE.CHECK]: ['--check'],

  [OPTION_TYPE.CONFIG_EMOJI]: ['--config-emoji', 'recommended-from-cli,ðŸš²'],

  [OPTION_TYPE.IGNORE_CONFIG]: [
    '--ignore-config',
    'ignoredConfigFromCli1',
    '--ignore-config',
    'ignoredConfigFromCli2',
  ],

  [OPTION_TYPE.IGNORE_DEPRECATED_RULES]: ['--ignore-deprecated-rules', 'true'],

  [OPTION_TYPE.INIT_RULE_DOCS]: ['--init-rule-docs', 'false'],

  [OPTION_TYPE.PATH_RULE_DOC]: [
    '--path-rule-doc',
    'www.example.com/rule-doc-from-cli',
  ],

  [OPTION_TYPE.PATH_RULE_LIST]: [
    '--path-rule-list',
    'www.example.com/rule-list-from-cli',
  ],

  [OPTION_TYPE.RULE_DOC_NOTICES]: ['--rule-doc-notices', 'type'],

  [OPTION_TYPE.RULE_DOC_SECTION_EXCLUDE]: [
    '--rule-doc-section-exclude',
    'excludedSectionFromCli1',
    '--rule-doc-section-exclude',
    'excludedSectionFromCli2',
  ],

  [OPTION_TYPE.RULE_DOC_SECTION_INCLUDE]: [
    '--rule-doc-section-include',
    'includedSectionFromCli1',
    '--rule-doc-section-include',
    'includedSectionFromCli2',
  ],

  [OPTION_TYPE.RULE_DOC_SECTION_OPTIONS]: [
    '--rule-doc-section-options',
    'true',
  ],

  [OPTION_TYPE.RULE_DOC_TITLE_FORMAT]: ['--rule-doc-title-format', 'name'],

  [OPTION_TYPE.RULE_LIST_COLUMNS]: ['--rule-list-columns', 'type'],

  [OPTION_TYPE.SPLIT_BY]: ['--split-by', 'meta.docs.foo-from-cli'],

  [OPTION_TYPE.URL_CONFIGS]: [
    '--url-configs',
    'https://example.com/configs-url-from-cli',
  ],

  [OPTION_TYPE.URL_RULE_DOC]: [
    '--url-rule-doc',
    'https://example.com/rule-doc-url-from-cli',
  ],
  [OPTION_TYPE.POSTPROCESS]: [],
};

describe('cli', function () {
  describe('no options', function () {
    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('all CLI options, no config file options', function () {
    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          ...Object.values(cliOptionsAll).flat(),
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('all config files options, no CLI options', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
          version: '1.0.0',
        }),

        '.eslint-doc-generatorrc.json': JSON.stringify(configFileOptionsAll),
      });
    });

    afterEach(function () {
      mockFs.restore();
    });

    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('all CLI options and all config files options', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
          version: '1.0.0',
        }),

        '.eslint-doc-generatorrc.json': JSON.stringify(configFileOptionsAll),
      });
    });

    afterEach(function () {
      mockFs.restore();
    });

    it('merges correctly, with CLI options taking precedence', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          ...Object.values(cliOptionsAll).flat(),
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('pathRuleList as array in config file and CLI', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
          version: '1.0.0',
        }),

        '.eslint-doc-generatorrc.json': JSON.stringify({
          pathRuleList: ['listFromConfigFile1.md', 'listFromConfigFile2.md'],
        }),
      });
    });

    afterEach(function () {
      mockFs.restore();
    });

    it('merges correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          '--path-rule-list',
          'listFromCli1.md',
          '--path-rule-list',
          'listFromCli2.md',
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('boolean option - false (explicit)', function () {
    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          '--ignore-deprecated-rules',
          'false',
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('boolean option - true (explicit)', function () {
    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          '--ignore-deprecated-rules',
          'true',
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('boolean option - true (implicit)', function () {
    it('is called correctly', async function () {
      const stub = sinon.stub().resolves();
      await run(
        [
          'node', // Path to node.
          'eslint-doc-generator.js', // Path to this binary.

          '--ignore-deprecated-rules',
        ],
        stub
      );
      expect(stub.callCount).toBe(1);
      expect(stub.firstCall.args).toMatchSnapshot();
    });
  });

  describe('missing package.json `version` field', function () {
    beforeEach(function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
        }),
      });
    });

    afterEach(function () {
      mockFs.restore();
    });

    it('throws an error', async function () {
      const stub = sinon.stub().resolves();
      await expect(
        run(
          [
            'node', // Path to node.
            'eslint-doc-generator.js', // Path to this binary.
          ],
          stub
        )
      ).rejects.toThrow('Could not find package.json `version`.');
    });
  });

  describe('invalid config file', function () {
    afterEach(function () {
      mockFs.restore();
    });

    it('throws an error', async function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
          version: '1.0.0',
        }),

        '.eslint-doc-generatorrc.json': JSON.stringify({
          // Doesn't match schema.
          unknown: true,
        }),
      });

      const stub = sinon.stub().resolves();
      await expect(
        run(
          [
            'node', // Path to node.
            'eslint-doc-generator.js', // Path to this binary.
          ],
          stub
        )
      ).rejects.toThrow('config file must NOT have additional properties');
    });

    it('requires that postprocess be a function', async function () {
      mockFs({
        'package.json': JSON.stringify({
          name: 'eslint-plugin-test',
          main: 'index.js',
          type: 'module',
          version: '1.0.0',
        }),

        '.eslint-doc-generatorrc.json': JSON.stringify({
          // Doesn't match schema.
          postprocess: './my-file.js',
        }),
      });

      const stub = sinon.stub().resolves();
      await expect(
        run(
          [
            'node', // Path to node.
            'eslint-doc-generator.js', // Path to this binary.
          ],
          stub
        )
      ).rejects.toThrow('postprocess must be a function');
    });
  });
});
