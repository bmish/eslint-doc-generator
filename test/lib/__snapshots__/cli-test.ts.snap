// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli all CLI options and all config files options merges correctly, with CLI options taking precedence 1`] = `
[
  ".",
  {
    "check": true,
    "configEmoji": [
      [
        "recommended-from-config-file",
        "ðŸš²",
      ],
      [
        "recommended-from-cli",
        "ðŸš²",
      ],
    ],
    "ignoreConfig": [
      "ignoredConfigFromConfigFile1",
      "ignoredConfigFromConfigFile2",
      "ignoredConfigFromCli1",
      "ignoredConfigFromCli2",
    ],
    "ignoreDeprecatedRules": true,
    "initRuleDocs": false,
    "pathRuleDoc": "www.example.com/rule-doc-from-cli",
    "pathRuleList": [
      "www.example.com/rule-list-from-config-file",
      "www.example.com/rule-list-from-cli",
    ],
    "ruleDocNotices": [
      "fixable",
      "type",
    ],
    "ruleDocSectionExclude": [
      "excludedSectionFromConfigFile1",
      "excludedSectionFromConfigFile2",
      "excludedSectionFromCli1",
      "excludedSectionFromCli2",
    ],
    "ruleDocSectionInclude": [
      "includedSectionFromConfigFile1",
      "includedSectionFromConfigFile2",
      "includedSectionFromCli1",
      "includedSectionFromCli2",
    ],
    "ruleDocSectionOptions": true,
    "ruleDocTitleFormat": "name",
    "ruleListColumns": [
      "fixable",
      "hasSuggestions",
      "type",
    ],
    "splitBy": "meta.docs.foo-from-cli",
    "urlConfigs": "https://example.com/configs-url-from-cli",
    "urlRuleDoc": "https://example.com/rule-doc-url-from-cli",
  },
]
`;

exports[`cli all CLI options, no config file options is called correctly 1`] = `
[
  ".",
  {
    "check": true,
    "configEmoji": [
      [
        "recommended-from-cli",
        "ðŸš²",
      ],
    ],
    "ignoreConfig": [
      "ignoredConfigFromCli1",
      "ignoredConfigFromCli2",
    ],
    "ignoreDeprecatedRules": true,
    "initRuleDocs": false,
    "pathRuleDoc": "www.example.com/rule-doc-from-cli",
    "pathRuleList": [
      "www.example.com/rule-list-from-cli",
    ],
    "ruleDocNotices": [
      "type",
    ],
    "ruleDocSectionExclude": [
      "excludedSectionFromCli1",
      "excludedSectionFromCli2",
    ],
    "ruleDocSectionInclude": [
      "includedSectionFromCli1",
      "includedSectionFromCli2",
    ],
    "ruleDocSectionOptions": true,
    "ruleDocTitleFormat": "name",
    "ruleListColumns": [
      "type",
    ],
    "splitBy": "meta.docs.foo-from-cli",
    "urlConfigs": "https://example.com/configs-url-from-cli",
    "urlRuleDoc": "https://example.com/rule-doc-url-from-cli",
  },
]
`;

exports[`cli all config files options, no CLI options is called correctly 1`] = `
[
  ".",
  {
    "check": true,
    "configEmoji": [
      [
        "recommended-from-config-file",
        "ðŸš²",
      ],
    ],
    "ignoreConfig": [
      "ignoredConfigFromConfigFile1",
      "ignoredConfigFromConfigFile2",
    ],
    "ignoreDeprecatedRules": true,
    "initRuleDocs": true,
    "pathRuleDoc": "www.example.com/rule-doc-from-config-file",
    "pathRuleList": [
      "www.example.com/rule-list-from-config-file",
    ],
    "ruleDocNotices": [
      "fixable",
    ],
    "ruleDocSectionExclude": [
      "excludedSectionFromConfigFile1",
      "excludedSectionFromConfigFile2",
    ],
    "ruleDocSectionInclude": [
      "includedSectionFromConfigFile1",
      "includedSectionFromConfigFile2",
    ],
    "ruleDocSectionOptions": false,
    "ruleDocTitleFormat": "desc",
    "ruleListColumns": [
      "fixable",
      "hasSuggestions",
    ],
    "splitBy": "meta.docs.foo-from-config-file",
    "urlConfigs": "https://example.com/configs-url-from-config-file",
    "urlRuleDoc": "https://example.com/rule-doc-url-from-config-file",
  },
]
`;

exports[`cli boolean option - false (explicit) is called correctly 1`] = `
[
  ".",
  {
    "configEmoji": [],
    "ignoreConfig": [],
    "ignoreDeprecatedRules": false,
    "pathRuleList": [],
    "ruleDocNotices": [],
    "ruleDocSectionExclude": [],
    "ruleDocSectionInclude": [],
    "ruleListColumns": [],
  },
]
`;

exports[`cli boolean option - true (explicit) is called correctly 1`] = `
[
  ".",
  {
    "configEmoji": [],
    "ignoreConfig": [],
    "ignoreDeprecatedRules": true,
    "pathRuleList": [],
    "ruleDocNotices": [],
    "ruleDocSectionExclude": [],
    "ruleDocSectionInclude": [],
    "ruleListColumns": [],
  },
]
`;

exports[`cli boolean option - true (implicit) is called correctly 1`] = `
[
  ".",
  {
    "configEmoji": [],
    "ignoreConfig": [],
    "ignoreDeprecatedRules": true,
    "pathRuleList": [],
    "ruleDocNotices": [],
    "ruleDocSectionExclude": [],
    "ruleDocSectionInclude": [],
    "ruleListColumns": [],
  },
]
`;

exports[`cli no options is called correctly 1`] = `
[
  ".",
  {
    "configEmoji": [],
    "ignoreConfig": [],
    "pathRuleList": [],
    "ruleDocNotices": [],
    "ruleDocSectionExclude": [],
    "ruleDocSectionInclude": [],
    "ruleListColumns": [],
  },
]
`;

exports[`cli pathRuleList as array in config file and CLI merges correctly 1`] = `
[
  ".",
  {
    "configEmoji": [],
    "ignoreConfig": [],
    "pathRuleList": [
      "listFromConfigFile1.md",
      "listFromConfigFile2.md",
      "listFromCli1.md",
      "listFromCli2.md",
    ],
    "ruleDocNotices": [],
    "ruleDocSectionExclude": [],
    "ruleDocSectionInclude": [],
    "ruleListColumns": [],
  },
]
`;
